scene/gui/text_edit.cpp

```diff
@@ -1083,11 +1071,31 @@ void TextEdit::_notification(int p_what) {
 						char_margin = size.width - char_margin - TS->shaped_text_get_size(rid).x;
 					}
 
+					float char_w = font->get_char_size(' ', font_size).width;
 					for (int c = 0; c < carets.size(); c++) {
 						if (!clipped && has_selection(c) && line >= get_selection_from_line(c) && line <= get_selection_to_line(c)) { // Selection
 							int sel_from = (line > get_selection_from_line(c)) ? TS->shaped_text_get_range(rid).x : get_selection_from_column(c);
 							int sel_to = (line < get_selection_to_line(c)) ? TS->shaped_text_get_range(rid).y : get_selection_to_column(c);
 							Vector<Vector2> sel = TS->shaped_text_get_selection(rid, sel_from, sel_to);
+							// End of line selected
+							if (line < get_selection_to_line(c)) {
+								Vector<Vector2> char_bounds = TS->shaped_text_get_selection(rid, 0, str.length());
+								float last_char_offset = 0;
+								Rect2 rect;
+								if (rtl) {
+									// Find the leftmost character's bounds
+									for (int j = 0; j < char_bounds.size(); j++) {
+										last_char_offset = MIN(last_char_offset, char_bounds[j].x);
```

Comment:
Won't this always be `0`?

Reply:
No, if you align text right, which makes sense in an rtl context. I don't know if there's any way to see this in code editor, but it works if you set TextEdit layout direction as rtl.
However, I just realized that you can use right justification for ltr text too, I should bind this to the `text_direction` property, not layout direction. In that case using left justified rtl text or right justified ltr text, the end of line won't be visible since there is no margin to draw them to, but I think it's a _marginal_ use case anyway 😅